lean4-infoview/src/infoview/userWidget.tsx

```diff
@@ -0,0 +1,125 @@
+import * as React from 'react';
+import type { Location } from 'vscode-languageserver-protocol';
+
+import { EditorContext, RpcContext } from './contexts';
+import { Details } from './collapsing';
+import { mapRpcError, } from './rpcInterface';
+import { DocumentPosition, useAsync, useEventResult } from './util';
+import { ErrorBoundary } from './errors';
+import { RpcSessions } from './rpcSessions';
+import { isRpcError, RpcErrorCode } from '@lean4/infoview-api';
+
+export interface GetWidgetResponse {
+    id: string
+    hash: number
+    props: any
+}
+
+function handleWidgetRpcError(e: unknown): undefined {
+    if (isRpcError(e)) {
+        if (e.code === RpcErrorCode.MethodNotFound || e.code === RpcErrorCode.InvalidParams) {
+            return undefined
+        } else {
+            throw Error(`RPC Error: ${RpcErrorCode[e.code]}: ${e.message}`)
+        }
+    } else if (e instanceof Error) {
+        throw e
+    } else {
+        throw Error(`Unknown rpc error ${JSON.stringify(e)}`)
+    }
+}
+
+export function Widget_getWidget(rs: RpcSessions, pos: DocumentPosition): Promise<GetWidgetResponse | undefined> {
+    return rs.call<GetWidgetResponse | undefined>(pos, 'Lean.Widget.getWidget', DocumentPosition.toTdpp(pos))
+        .catch<undefined>(handleWidgetRpcError);
+}
+
+export interface StaticJS {
+    javascript: string
+    hash: number
+}
+
+/** Gets the static JS code for a given widget.
+ *
+ * We make the assumption that either the code doesn't exist, or it exists and does not change for the lifetime of the widget.
+ * [todo] cache on widgetId, but then there needs to be some way of signalling that the widgetId's code has changed if the user edits it?
+ */
+export async function Widget_getStaticJS(rs: RpcSessions, pos: DocumentPosition, widgetId: string): Promise<StaticJS | undefined> {
```

Comment:
```suggestion
export async function Widget_getStaticJS(rs: RpcSessions, pos: DocumentPosition, hash: number): Promise<StaticJS | undefined> {
```
Just treat the server as a content-addressed javascript CDN.  Then it doesn't matter if the code changes while the extension is running.